rho/utilities.py

```diff
@@ -243,19 +252,120 @@ def check_range_validity(range_list):
                   '[a-zA-Z0-9-\.]*\[[0-9]*:[0-9]*\]*[a-zA-Z0-9-\.]*',
                   '[a-zA-Z0-9-\.]*\[[a-zA-Z]*:[a-zA-Z]*\][a-zA-Z0-9-\.]*']
 
+    if ranges_or_path and os.path.isfile(ranges_or_path[0]):
+        range_list = _read_in_file(ranges_or_path[0])
+    else:
+        range_list = ranges_or_path
+
+    normalized = []
+
     for reg_item in range_list:
-        match = False
+        match_found = False
         for reg in regex_list:
-            matches = re.findall(reg, reg_item)
-            if len(matches) == 1:
-                match = True
-        if not match:
-            if len(reg_item) <= 1:
-                print(_("No such hosts file."))
-
-            print(_("Bad host name/range : '%s'") % reg_item)
+            match = re.match(reg, reg_item)
+            if match and match.end() == len(reg_item):
+                normalized.append(reg_item)
+                match_found = True
+                break
+
+        if not match_found:
+            try:
+                normalized.append(cidr_to_ansible(reg_item))
+                match_found = True
+            except NotCIDRException:
+                pass
+
+        if not match_found:
+            logging.error(_("Bad host name/range : '%s'"), reg_item)
+            if len(range_list) <= 1:
+                logging.error(_("Couldn't interpret %s as host file because "
+                                "no such file", reg_item))
+
+            sys.exit(1)
+
+    return normalized
+
+
+class NotCIDRException(Exception):
+    """Exception for when a string does not look like a CIDR range."""
+
+    pass
+
+
+def cidr_to_ansible(ip_range):
+    """Convert an IP address range from CIDR to Ansible notation.
+
+    :param ip_range: the IP range, as a string
+    :returns: the IP range, as an Ansible-formatted string
+
+    Raises NotCIDRException if ip_range doesn't look similar to CIDR
+    notation. If it does look like CIDR but isn't quite right, print
+    out error messages and exit.
+    """
+
+    # In the case of an input error, we want to distinguish between
+    # strings that are "CIDR-like", so the user probably intended to
+    # use CIDR and we should give them a CIDR error message, and not
+    # at all CIDR-like, in which case we tell the caller to parse it a
+    # different way.
+    cidr_like = r'[0-9.]*/[0-9]*'
+
+    if not re.match(cidr_like, ip_range):
+        raise NotCIDRException
+
+    base_address, prefix_bits = ip_range.split('/')
+    prefix_bits = int(prefix_bits)
+
+    if prefix_bits < 0 or prefix_bits > 32:
+        logging.error('Bit mask length %s of IP range %s is not in the valid '
+                      'range [0,32].', prefix_bits, ip_range)
+        sys.exit(1)
```

Comment:
Seems like we are exiting here without outputting an error to the user. Its just being logged.

Reply:
You're right. What do you think of the idea of changing our logging setup so that all messages go to the Rho log, and messages at WARNING level or higher are sent to stderr? I just took a look at the logging module, and it seems like that's not too hard to do.

Reply:
I theory it sounds good, but I think we would want to also "scope" our logger so its no longer global( #256 ) otherwise we don't know what might bubble up to the user from other modules that may cause some confusion.